daemon/state.go

```diff
@@ -97,13 +98,14 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
 			endpointmanager.Insert(ep)
 			epRestored <- true
 
+			ep.LogStatusOKLocked(endpoint.Other, "Synchronizing endpoint labels with KVStore")
 			if err := d.syncLabels(ep); err != nil {
 				scopedLog.WithError(err).Warn("Unable to restore endpoint")
 				ep.Mutex.Unlock()
 				epRegenerated <- false
 				return
 			}
-			ready := ep.SetStateLocked(endpoint.StateRestoring, "Triggering synchronous endpoint regeneration while syncing state to host")
+			ready := ep.SetStateLocked(endpoint.StateWaitingToRegenerate, "Triggering synchronous endpoint regeneration while syncing state to host")
```

Comment:
I'm late in the game but is this change needed given that you also changed regenerateBPF() to recognize `StateRestoring`? Or the other way around, when this change is made, do we ever call regenerateBPF() while the endpoint's state is `StateRestoring`?

Reply:
(I'm not aanm but we had chatted about this)
I think the idea was to more consistently enter `WaitingToRegenerate` and then `Regenerating`. While `regenerateBPF` does try to switch to `Regenerating` even if the state is `Restoring` (because we removed the `!=` check on line 425 in `regenerateBPF`) the state machine itself in `BuilderSetState` won't allow that transition. With this change the endpoint follows the same state progression it would if it was being normally created, or had a regeneration triggered for any other reason (there's a nice cycle of Ready -> WaitingToRegenerate -> Regenerating -> Ready that seems to be our normal operational state).
So, to answer your actual question: it's needed unless we change the state machine. This seemed like the simpler option.
On whether we call `regenerateBPF` while `Restoring`: I don't know. Hopefully not but it would then trigger an error. I would argue this is desirable since we can then fix it but there are many callpaths so that might be a lot of work :(

Reply:
@jrajahalme IIRC regenerateBPF no longer recognizes `StateRestoring`